test/tests.js

```diff
@@ -57,6 +58,112 @@ describe('Dylan', function() {
     });
   });
 
+  it('returns 404s for unhandled requests', (done) => {
+    app.listen(8888);
+
+    chai.request(host)
+      .get('/invalid-path')
+      .end((err, res) => {
+        expect(res.status).to.equal(404);
+        expect(res.text).to.include('Not Found');
+        done();
+      });
+  });
+
+  describe('error handler', function() {
+    it('sends a default error response on error', (done) => {
+      app.get('/error', (req, res) => {
+        throw "Error!";
+      });
+      app.listen(8888);
+
+      chai.request(host)
+        .get('/error')
+        .end((err, res) => {
+          expect(res.status).to.equal(500);
+          expect(res.text).to.include('500');
+          done();
+        });
+    });
+
+    describe('in production', function() {
+      let prevEnv;
+
+      beforeEach(() => {
+        prevEnv = process.env.NODE_ENV;
+        process.env.NODE_ENV = 'production';
+      });
+
+      afterEach(() => process.env.NODE_ENV = prevEnv);
+
+      it('sends a default error response on error', (done) => {
+        app.get('/error', (req, res) => {
+          throw "Error!";
+        });
+        app.listen(8888);
+
+        chai.request(host)
+          .get('/error')
+          .end((err, res) => {
+            expect(res.status).to.equal(500);
+            expect(res.text).to.include('500');
+            done();
+          });
+      });
+    });
+
+    describe('configured', function() {
+      beforeEach(() => {
+        config.errorHandle = (err, req, res, next) => {
+          res.end('error!');
+        };
+      });
+
+      afterEach(() => delete config.errorHandle);
+
+      it('uses the error handler on error', (done) => {
+        app.get('/error', (req, res) => {
+          throw "Error!";
+        });
+        app.listen(8888);
+
+        chai.request(host)
+          .get('/error')
+          .end((err, res) => {
+            expect(res.status).to.equal(500);
+            expect(res.text).to.include('error!');
+            done();
+          });
+      });
+
+      describe('in production', function() {
+        let prevEnv;
+
+        beforeEach(() => {
+          prevEnv = process.env.NODE_ENV;
+          process.env.NODE_ENV = 'production';
+        });
+
+        afterEach(() => process.env.NODE_ENV = prevEnv);
```

Comment:
This code is quite confusing. What's it doing?

Reply:
`beforeEach` is a function that runs before each test runs.

`afterEach` is a function that runs after each test runs.

The `beforeEach` function has two lines of code:
```
          prevEnv = process.env.NODE_ENV;
          process.env.NODE_ENV = 'production';
```

The first line is assigning the current value of `NODE_ENV` to `prevEnv`.
The second line is setting `NODE_ENV` to `production`.

The `afterEach` function is restoring the value of `NODE_ENV` to `prevEnv`